File: src/core/expr/code/freevar.rb
require 'core/expr/code/eval'
require 'core/expr/code/lvalue'
require 'core/semantics/code/interpreter'

module Freevar
  module FreeVarExpr
    #Determine the set of unbounded variables in this expr
    #that need to be supplied by the environment.
    #There are no assumptions as to the type of EFields,
    #so they can be MObjects or Ruby objects or even arrays
  
    #Function calls do not work properly here since functions
    #can be defined outside of the expression (they are ALWAYS
    #defined externally if this is Expr without Impl)
  
    include Eval::EvalExpr
    include Lvalue::LValueExpr
    
    include Interpreter::Dispatcher    
      
    def depends(obj)
      dispatch_obj(:depends, obj)
    end
    
    def depends_EField(obj)
      #[*depends(obj.e)]
      depends(obj.e) 
    end
  
    def depends_EVar(obj)
      (@D[:bound].include?(obj.name) || obj.name == "self") ? [] : [Lvalue::Address.new(@D[:env], obj.name)]
    end
  
    def depends_ELambda(obj)
      bound2 = @D[:bound].clone
      obj.formals.each{|f|bound2<<depends(f)}
      dynamic_bind(bound: bound2) do
        depends(obj.body)
      end
    end
    
    def depends_Formal(obj)
      obj.name
    end
  
    def depends_?(obj)
      res = []
      type = obj.schema_class
      type.fields.each do |f|
        if f.traversal && !f.type.Primitive? && obj[f.name]
          if !f.many
            res = res.concat( depends(obj[f.name]) )
          else
            obj[f.name].each {|o| res = res.concat(depends(o)) }
          end
        end
      end
      res
    end
  end
  
  class FreeVarExprC
    include FreeVarExpr  
    def initialize
    end
  end

  def self.depends(obj, args={})
    interp = FreeVarExprC.new
    interp.dynamic_bind args do
      interp.depends(obj)
    end
  end
end

File: src/js/core/expr/code/freevar.js
'use strict'

//// Freevar ////

var cwd = process.cwd() + '/';
var Eval = require(cwd + "core/expr/code/eval.js");
var Lvalue = require(cwd + "core/expr/code/lvalue.js");
var Interpreter = require(cwd + "core/semantics/code/interpreter.js");
var Enso = require(cwd + "enso.js");

var Freevar;

var depends = function(obj, args = Enso.EMap.new()) {
  var self = this, interp;
  interp = FreeVarExprC.new();
  return interp.dynamic_bind(function() {
    return interp.depends(obj);
  }, args);
};

function FreeVarExpr(parent) {
  return class extends Enso.mix(parent, Eval.EvalExpr, Lvalue.LValueExpr, Interpreter.Dispatcher) {
    depends(obj) {
      var self = this;
      return self.dispatch_obj("depends", obj);
    };

    depends_EField(obj) {
      var self = this;
      return self.depends(obj.e());
    };

    depends_EVar(obj) {
      var self = this;
      if (self.D$.get$("bound").include_P(obj.name()) || obj.name() == "self") {
        return [];
      } else {
        return [Lvalue.Address.new(self.D$.get$("env"), obj.name())];
      }
    };

    depends_ELambda(obj) {
      var self = this, bound2;
      bound2 = self.D$.get$("bound").clone();
      obj.formals().each(function(f) {
        return bound2.push(self.depends(f));
      });
      return self.dynamic_bind(function() {
        return self.depends(obj.body());
      }, Enso.EMap.new({bound: bound2}));
    };

    depends_Formal(obj) {
      var self = this;
      return obj.name();
    };

    depends__P(obj) {
      var self = this, res, type;
      res = [];
      type = obj.schema_class();
      type.fields().each(function(f) {
        if ((f.traversal() && ! Enso.System.test_type(f.type(), "Primitive")) && obj.get$(f.name())) {
          if (! f.many()) {
            return res = res.concat(self.depends(obj.get$(f.name())));
          } else {
            return obj.get$(f.name()).each(function(o) {
              return res = res.concat(self.depends(o));
            });
          }
        }
      });
      return res;
    }; }};

class FreeVarExprC extends Enso.mix(Enso.EnsoBaseClass, FreeVarExpr) {
  static new(...args) { return new FreeVarExprC(...args) };

  constructor() {
    super();
  };
};

Freevar = {
  depends: depends,
  FreeVarExpr: FreeVarExpr,
  FreeVarExprC: FreeVarExprC,
};
module.exports = Freevar ;
